### ISSUE TITLE
Option to ignore no-nested-ternary for daisy-chained ternaries only

### ISSUE THREAD

**Issue by zeorin** [eslint/eslint#10321]
**What rule do you want to change?**
`no-nested-ternary`

**Does this change cause the rule to produce more or fewer warnings?**
Fewer warnings.

**How will the change be implemented? (New option, new default behavior, etc.)?**
New option, e.g.: `allowDaisyChain: true`

**Please provide some example code that this change will affect:**

```js
var foo = false;
var bar = true;

var baz =
	foo ? 'foo' :
	bar ? 'bar' :
	null;

var qux =
	!foo ?
		!bar ? null : 'bar'
	: 'foo';
```
([ESLint Demo](https://eslint.org/demo/#eyJ0ZXh0IjoidmFyIGZvbyA9IGZhbHNlO1xudmFyIGJhciA9IHRydWU7XG5cbnZhciBiYXogPVxuXHRmb28gPyAnZm9vJyA6XG5cdGJhciA/ICdiYXInIDpcblx0bnVsbDtcblxudmFyIHF1eCA9XG5cdCFmb28gP1xuXHRcdCFiYXIgPyBudWxsIDogJ2Jhcidcblx0OiAnZm9vJzsiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjUsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsibm8tbmVzdGVkLXRlcm5hcnkiOjJ9LCJlbnYiOnt9fX0=))
Both `baz` and `qux` will be `'bar'`, but the code for the assignment to `baz` is far more easy to reason about because it's daisy-chained and not just nested.

It'd be great to allow nested ternaries of that type whilst still disallowing the more convoluted `qux` form.

**What does the rule currently do for this code?**
Raises 2 errors:
> 5:2 - Do not nest ternary expressions. (no-nested-ternary)
> 10:2 - Do not nest ternary expressions. (no-nested-ternary)

**What will the rule do after it's changed?**
Raise only one error:
> 10:2 - Do not nest ternary expressions. (no-nested-ternary)

For a more in-depth analysis of nested ternaries and how the daisy-chained version is useful because it's like a switch statement turned into an expression, capable of more than just strict equality checks of one value, see here: https://medium.com/javascript-scene/nested-ternaries-are-great-361bddd0f340#c69a

**Comment by platinumazure** [comment]
Hi @zeorin, thanks for the issue!

Personally, I'm having a hard time seeing the benefit compared to the amount of logic we would need to add to the rule (and maintain). I think this would not be a large benefit for many users. So I'm :-1: to adding this to the core rule.

If this enhancement isn't accepted by the team, you are welcome to create a custom version of the rule with the desired behavior.

**Comment by zeorin** [comment]
@platinumazure, I added a pull request implementing the change. It was rather simple to implement. :smile:

I understand it seems like a bit of an edge case, but with functional programming becoming more popular, it's likely that there will be an uptake in conditional _expressions_ vs. _statements_, and a daisy-chained ternary is a great way to turn a switch statement into an expression.

:crossed_fingers: hoping that will encourage adding the option.

**Comment by platinumazure** [comment]
Thanks @zeorin, that helps a bit. I'm now willing to support the option.

I'd like to suggest that the option name should be changed. "Daisy chain" is an idiom that non-English speakers will struggle to understand. I think something like `allowIfAlternate`, `allowAlternate`, or `allowConsecutiveAlternateTernary` (or similar) would work just as well with less risk of confusing a non-English speaker. What do you think?

**Comment by zeorin** [comment]
That's a good point. However, I only came across the terms "consequent" and "alternate" in the documentation for AST. To me, "alternate" isn't very clear, either.

I did check [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Conditional_Operator) for the terms they used, but `expr1` and `expr2` aren't any better.

How about `allowTailNesting`? (or is that confusing w.r.t. "tail call"?)

**Comment by platinumazure** [comment]
I could also support something like `allowConsecutive` if that isn't judged to be too ambiguous. Agreed that `allowTailNesting` could be confusing and definitely agree `expr1` and `expr2` are rubbish. :smile:

**Comment by zeorin** [comment]
`allowElseTernary`?

**Comment by zeorin** [comment]
@platinumazure I updated the pull request: `allowDaisyChain` is now `allowAlternate`, and I added a short explanation to the docs of the meaning of "alternate" in this context.

**Comment by whitelizard** [comment]
I would love this option.

I have missed to be able to do this for a long time (with still good linting active + integration with _prettier_ formatting, etc - i.e. an exception option to the no-nested-ternary rule).

I'm thinking about this article on the subject that argues for the use of chained ternaries, and that I agree on:
[https://medium.com/javascript-scene/nested-ternaries-are-great-361bddd0f340](https://medium.com/javascript-scene/nested-ternaries-are-great-361bddd0f340)

Also as @zeorin said:
> with functional programming becoming more popular, it's likely that there will be an uptake in conditional expressions vs. statements

Yes.
